modin/data_management/functions/default_methods/default.py

```diff
@@ -0,0 +1,100 @@
+# Licensed to Modin Development Team under one or more contributor license agreements.
+# See the NOTICE file distributed with this work for additional information regarding
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+from modin.data_management.functions.function import Function
+from modin.utils import try_cast_to_pandas
+
+from pandas.core.dtypes.common import is_list_like
+import pandas
+
+
+class DefaultMethod(Function):
+    @classmethod
+    def call(cls, func, **call_kwds):
+        obj = call_kwds.get("obj_type", pandas.DataFrame)
+        force_inplace = call_kwds.get("inplace")
+
+        if isinstance(func, str):
+            fn = getattr(obj, func)
+        else:
+            fn = func
+
+        if type(fn) == property:
+            fn = cls.build_property_wrapper(fn)
+
+        def applyier(df, *args, **kwargs):
+            df = cls.frame_wrapper(df)
+            result = fn(df, *args, **kwargs)
+
+            if (
+                not isinstance(result, pandas.Series)
+                and not isinstance(result, pandas.DataFrame)
+                and func != "to_numpy"
+                and func != pandas.DataFrame.to_numpy
+            ):
+                result = (
+                    pandas.DataFrame(result)
+                    if is_list_like(result)
+                    else pandas.DataFrame([result])
+                )
+            if isinstance(result, pandas.Series):
+                if result.name is None:
+                    result.name = "__reduced__"
+                result = result.to_frame()
+
+            inplace = kwargs.get("inplace", False)
+            if force_inplace is not None:
+                inplace = force_inplace
+            return result if not inplace else df
+
+        return cls.build_wrapper(applyier, func)
+
+    @classmethod
+    def register(cls, func, **kwargs):
```

Comment:
I know it's a bit late to voice questions on a merged PR, but I had this open for review yesterday and wasn't able to comment on.

I don't think the naming is correct, as `.register()` usually imples a one-time action that creates a record, but here it actually is used to call a thing.
I'm also slightly worried about performance, as each Python call is somewhat expensieve...

Reply:
it's better late, than never)

`.register()` is used to create default to pandas functions, yes, in [some cases](https://github.com/modin-project/modin/blob/90adc7d7cbbb424c27f0618b94cb006af06dc5e7/modin/backends/base/query_compiler.py#L414), when a method in `BaseQueryCompiler` has its interface, we're calling the result of that function in place, but there is also [a lot of examples](https://github.com/modin-project/modin/blob/90adc7d7cbbb424c27f0618b94cb006af06dc5e7/modin/backends/base/query_compiler.py#L1531-L1733) when we actually register function with that method. We can add an additional method to `DefaultMethod` called somehow as `apply` that will do `.register()()`:
```python
DataFrameDefault.apply(pandas.DataFrame.method, self, *args, **kwargs)
```
Will that be okay?


About a lot of function calls and performance. If we will call the result of `.register()` function, we will get a call stack that looks something like this:
1. [Function](https://github.com/modin-project/modin/blob/90adc7d7cbbb424c27f0618b94cb006af06dc5e7/modin/data_management/functions/default_methods/default.py#L93-L94) that wraps default_to_pandas call
2. `default_to_pandas()`
3. [`args_cast()`](https://github.com/modin-project/modin/blob/90adc7d7cbbb424c27f0618b94cb006af06dc5e7/modin/data_management/functions/default_methods/default.py#L70-L73) - casting function arguments to pandas
4. [Function](https://github.com/modin-project/modin/blob/90adc7d7cbbb424c27f0618b94cb006af06dc5e7/modin/data_management/functions/default_methods/default.py#L86-L87) that wraps call of `fget()` if property
5. [`applier()`](https://github.com/modin-project/modin/blob/90adc7d7cbbb424c27f0618b94cb006af06dc5e7/modin/data_management/functions/default_methods/default.py#L35-L58) - wrapping the result of the applied function (casting Series, arrays, scalars... to DataFrame)
6. [`.frame_wrapper()`](https://github.com/modin-project/modin/blob/90adc7d7cbbb424c27f0618b94cb006af06dc5e7/modin/data_management/functions/default_methods/default.py#L36) - squeeze DataFrame or get some property of frame on which we want to apply our function
7. the function itself


We usually want to overwrite some of these steps, depends on what function we're building.
For binary functions, we want to [overwrite](https://github.com/modin-project/modin/blob/90adc7d7cbbb424c27f0618b94cb006af06dc5e7/modin/data_management/functions/default_methods/binary_default.py#L23-L42) 3rd step, because in binary functions QC obtains arguments, that tell him to squeeze `self`, `other`, or both. So we want to grab these args at this stage and do what it needs to do depends on those arguments.

There is always a trade-off between performance and usability of code, but I'm wondering, will it be some noticeable overhead of calling +6 extra functions when `default_to_pandas` is already a pretty slow operation? It must grab DataFrames from every partition, combine them together, apply a function to that frame, and then split it and put it into partitions again. How much expensive a Python function call is?